util/upload-cookbook.py

```diff
@@ -0,0 +1,241 @@
+
+            assumed_role_object = sts.assume_role(
+                RoleArn=credential_arn,
+                ExternalId=credential_external_id,
+                RoleSessionName=credential_region + "upload_cfn_templates_sts_session",
+            )
+            aws_credentials = assumed_role_object["Credentials"]
+            s3 = boto3.client(
+                "s3",
+                region_name=credential_region,
+                aws_access_key_id=aws_credentials.get("AccessKeyId"),
+                aws_secret_access_key=aws_credentials.get("SecretAccessKey"),
+                aws_session_token=aws_credentials.get("SessionToken"),
+            )
+
+        except ClientError:
+            print("Warning: non authorized in region '{0}', skipping".format(credential_region))
+            pass
+    else:
+        s3 = boto3.client("s3")
+    return s3
+
+
+def _get_bucket_name(args, region):
+    return region + "-aws-parallelcluster" if not args.bucket else args.bucket
+
+
+def main(args):
+
+    for _region in args.regions:
+        s3 = _create_s3_client(_region)
+        bucket1_name = _get_bucket_name(args, _region)
+
+        print("Listing cookbook for region ({0})".format(_region))
+        _aws_ls(s3, _region, bucket1_name, _COOKBOOKS_DIR + "/" + args.full_name + ".tgz")
+
+    if len(_ls_error_array) > 0 and not args.override:
+        print("We know the cookbook archives are already there, in this round we need to upload the .date files!")
+        print("Failed to push cookbook, already present for region " + " ".join(_ls_error_array))
+        exit(1)
+    elif len(_ls_error_array) > 0 and args.override:
+        print("Some or all of the cookbook archives are already there but OVERRIDE=true")
+
+    for _region in args.regions:
+        s3 = _create_s3_client(_region)
+        bucket1_name = _get_bucket_name(args, _region)
+
+        if args.override:
+            print("Backup cookbook for region :{0}".format(_region))
+            _aws_bck(s3, args, _region, bucket1_name, args.full_name + ".tgz")
+            _aws_bck(s3, args, _region, bucket1_name, args.full_name + ".md5")
+            _aws_bck(s3, args, _region, bucket1_name, args.full_name + ".tgz.date")
+
+        print("Pushing cookbook for region: {0}".format(_region))
+        _aws_cp(s3, args, _region, bucket1_name, _COOKBOOKS_DIR, args.full_name + ".tgz")
+        _aws_cp(s3, args, _region, bucket1_name, _COOKBOOKS_DIR, args.full_name + ".md5")
+
+        if not args.dryrun:
+            # Stores LastModified info into .tgz.date file and uploads it back to bucket
+            with (open(args.full_name + ".tgz.date", "w+")) as f:
+                response = s3.head_object(Bucket=bucket1_name, Key=_COOKBOOKS_DIR + "/" + args.full_name + ".tgz")
+                f.write(response.get("LastModified").strftime("%Y-%m-%d_%H-%M-%S"))
+
+            _aws_cp(s3, args, _region, bucket1_name, _COOKBOOKS_DIR, args.full_name + ".tgz.date")
```

Comment:
I'd rename _aws_cp to copy_to_s3 or something more explicit

Reply:
Good idea, renaming also other functions to mimic aws' command s3 syntax, like below:

- _aws_s3_cp
- _aws_s3_bck
- _aws_s3_ls